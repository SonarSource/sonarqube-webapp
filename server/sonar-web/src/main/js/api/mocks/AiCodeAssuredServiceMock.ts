/*
 * SonarQube
 * Copyright (C) 2009-2024 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

import {
  AiCodeAssuranceStatus,
  getProjectBranchesAiCodeAssuranceStatus,
  getProjectContainsAiCode,
  getProjectDetectedAiCode,
} from '../ai-code-assurance';

jest.mock('../ai-code-assurance');

export const PROJECT_WITH_AI_ASSURED_QG = 'Sonar AI way';
export const PROJECT_WITHOUT_AI_ASSURED_QG = 'Sonar way';

export class AiCodeAssuredServiceMock {
  noAiProject = 'no-ai';

  defaultAIStatus = AiCodeAssuranceStatus.NONE;
  detectedAiCode = true;

  constructor() {
    jest
      .mocked(getProjectBranchesAiCodeAssuranceStatus)
      .mockImplementation(this.handleProjectBranchAiAssuranceStatus);

    jest.mocked(getProjectContainsAiCode).mockImplementation(this.handleProjectAiContainsCode);
    jest.mocked(getProjectDetectedAiCode).mockImplementation(this.handleProjectDetectedAiCode);
  }

  handleProjectBranchAiAssuranceStatus = (project: string) => {
    if (project === PROJECT_WITH_AI_ASSURED_QG) {
      return Promise.resolve(AiCodeAssuranceStatus.AI_CODE_ASSURED_ON);
    } else if (project === PROJECT_WITHOUT_AI_ASSURED_QG) {
      return Promise.resolve(AiCodeAssuranceStatus.AI_CODE_ASSURED_OFF);
    }
    return Promise.resolve(this.defaultAIStatus);
  };

  handleProjectAiContainsCode = (project: string) => {
    if (project === PROJECT_WITH_AI_ASSURED_QG || project === PROJECT_WITHOUT_AI_ASSURED_QG) {
      return Promise.resolve(true);
    }
    return Promise.resolve(false);
  };

  handleProjectDetectedAiCode = () => {
    return Promise.resolve(this.detectedAiCode);
  };

  reset() {
    this.noAiProject = 'no-ai';
    this.defaultAIStatus = AiCodeAssuranceStatus.NONE;
    this.detectedAiCode = true;
  }
}
