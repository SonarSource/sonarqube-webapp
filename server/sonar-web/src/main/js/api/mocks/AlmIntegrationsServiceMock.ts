/*
 * SonarQube
 * Copyright (C) 2009-2022 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { cloneDeep } from 'lodash';
import {
  mockAzureProject,
  mockAzureRepository,
  mockGitlabProject,
} from '../../helpers/mocks/alm-integrations';
import { AzureProject, AzureRepository, GitlabProject } from '../../types/alm-integration';
import {
  checkPersonalAccessTokenIsValid,
  getAzureProjects,
  getAzureRepositories,
  getGithubClientId,
  getGithubOrganizations,
  getGitlabProjects,
  importAzureRepository,
  searchAzureRepositories,
  setAlmPersonalAccessToken,
} from '../alm-integrations';
import { ProjectBase } from '../components';

export default class AlmIntegrationsServiceMock {
  almInstancePATMap: { [key: string]: boolean } = {};
  gitlabProjects: GitlabProject[];
  azureProjects: AzureProject[];
  azureRepositories: AzureRepository[];
  defaultAlmInstancePATMap: { [key: string]: boolean } = {
    'conf-final-1': false,
    'conf-final-2': true,
    'conf-github-1': false,
    'conf-github-2': true,
    'conf-azure-1': false,
    'conf-azure-2': true,
    'config-reject': false,
  };

  defaultGitlabProjects: GitlabProject[] = [
    mockGitlabProject({
      name: 'Gitlab project 1',
      id: '1',
      sqProjectKey: 'key',
      sqProjectName: 'Gitlab project 1',
    }),
    mockGitlabProject({ name: 'Gitlab project 2', id: '2' }),
    mockGitlabProject({ name: 'Gitlab project 3', id: '3' }),
  ];

  defaultAzureProjects: AzureProject[] = [
    mockAzureProject({ name: 'Azure project', description: 'Description project 1' }),
    mockAzureProject({ name: 'Azure project 2', description: 'Description project 2' }),
  ];

  defaultAzureRepositories: AzureRepository[] = [
    mockAzureRepository({ sqProjectKey: 'random' }),
    mockAzureRepository({ name: 'Azure repo 2' }),
  ];

  defaultOrganizations = {
    paging: {
      pageIndex: 1,
      pageSize: 100,
      total: 1,
    },
    organizations: [
      {
        key: 'org-1',
        name: 'org-1',
      },
    ],
  };

  constructor() {
    this.almInstancePATMap = cloneDeep(this.defaultAlmInstancePATMap);
    this.gitlabProjects = cloneDeep(this.defaultGitlabProjects);
    this.azureProjects = cloneDeep(this.defaultAzureProjects);
    this.azureRepositories = cloneDeep(this.defaultAzureRepositories);
    (checkPersonalAccessTokenIsValid as jest.Mock).mockImplementation(
      this.checkPersonalAccessTokenIsValid
    );
    (setAlmPersonalAccessToken as jest.Mock).mockImplementation(this.setAlmPersonalAccessToken);
    (getGitlabProjects as jest.Mock).mockImplementation(this.getGitlabProjects);
    (getGithubClientId as jest.Mock).mockImplementation(this.getGithubClientId);
    (getGithubOrganizations as jest.Mock).mockImplementation(this.getGithubOrganizations);
    (getAzureProjects as jest.Mock).mockImplementation(this.getAzureProjects);
    (getAzureRepositories as jest.Mock).mockImplementation(this.getAzureRepositories);
    (searchAzureRepositories as jest.Mock).mockImplementation(this.searchAzureRepositories);
    (importAzureRepository as jest.Mock).mockImplementation(this.importAzureRepository);
  }

  checkPersonalAccessTokenIsValid = (conf: string) => {
    return Promise.resolve({ status: this.almInstancePATMap[conf] });
  };

  setAlmPersonalAccessToken = (conf: string) => {
    this.almInstancePATMap[conf] = true;
    return Promise.resolve();
  };

  getAzureProjects = (): Promise<{ projects: AzureProject[] }> => {
    return Promise.resolve({ projects: this.azureProjects });
  };

  getAzureRepositories = (): Promise<{ repositories: AzureRepository[] }> => {
    return Promise.resolve({
      repositories: this.azureRepositories,
    });
  };

  searchAzureRepositories = (): Promise<{ repositories: AzureRepository[] }> => {
    return Promise.resolve({
      repositories: this.azureRepositories,
    });
  };

  setSearchAzureRepositories = (azureRepositories: AzureRepository[]) => {
    this.azureRepositories = azureRepositories;
  };

  importAzureRepository = (): Promise<{ project: ProjectBase }> => {
    return Promise.resolve({
      project: {
        key: 'key',
        name: 'name',
        qualifier: 'qualifier',
        visibility: 'private',
      },
    });
  };

  setAzureProjects = (azureProjects: AzureProject[]) => {
    this.azureProjects = azureProjects;
  };

  getGitlabProjects = () => {
    return Promise.resolve({
      projects: this.gitlabProjects,
      projectsPaging: {
        pageIndex: 1,
        pageSize: 30,
        total: this.gitlabProjects.length,
      },
    });
  };

  setGitlabProjects(gitlabProjects: GitlabProject[]) {
    this.gitlabProjects = gitlabProjects;
  }

  getGithubClientId = () => {
    return Promise.resolve({ clientId: 'clientId' });
  };

  getGithubOrganizations = () => {
    return Promise.resolve(this.defaultOrganizations);
  };

  reset = () => {
    this.almInstancePATMap = cloneDeep(this.defaultAlmInstancePATMap);
    this.gitlabProjects = cloneDeep(this.defaultGitlabProjects);
    this.azureRepositories = cloneDeep(this.defaultAzureRepositories);
  };
}
