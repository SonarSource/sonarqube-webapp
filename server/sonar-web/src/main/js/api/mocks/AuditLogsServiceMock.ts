/*
 * SonarQube
 * Copyright (C) 2009-2022 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { cloneDeep } from 'lodash';
import { HousekeepingPolicy } from '../../apps/audit-logs/utils';
import { SettingValue } from '../../types/settings';
import { getValues } from '../settings';

export default class AuditLogsServiceMock {
  settingValue: SettingValue[];
  defaultValues: SettingValue[] = [{ key: 'test', value: HousekeepingPolicy.Weekly }];

  constructor() {
    this.settingValue = cloneDeep(this.defaultValues);
    (getValues as jest.Mock).mockImplementation(this.getValuesHandler);
  }

  getValuesHandler = () => {
    return Promise.resolve(this.settingValue);
  };

  resetSettingvalues = () => {
    this.settingValue = cloneDeep(this.defaultValues);
  };
}
