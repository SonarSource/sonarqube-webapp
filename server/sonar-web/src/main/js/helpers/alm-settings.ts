/*
 * SonarQube
 * Copyright (C) 2009-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import {
  AlmBindingDefinition,
  AlmKeys,
  BitbucketBindingDefinition,
  GithubBindingDefinition,
  ProjectAlmBindingResponse,
  ProjectBitbucketBindingResponse,
  ProjectGitHubBindingResponse,
  ProjectGitLabBindingResponse
} from '../types/alm-settings';

export function isProjectBitbucketBindingResponse(
  binding: ProjectAlmBindingResponse
): binding is ProjectBitbucketBindingResponse {
  return binding.alm === AlmKeys.Bitbucket;
}

export function isProjectGitHubBindingResponse(
  binding: ProjectAlmBindingResponse
): binding is ProjectGitHubBindingResponse {
  return binding.alm === AlmKeys.GitHub;
}

export function isProjectGitLabBindingResponse(
  binding: ProjectAlmBindingResponse
): binding is ProjectGitLabBindingResponse {
  return binding.alm === AlmKeys.GitLab;
}

export function isBitbucketBindingDefinition(
  binding?: AlmBindingDefinition & { url?: string; personalAccessToken?: string }
): binding is BitbucketBindingDefinition {
  return (
    binding !== undefined && binding.url !== undefined && binding.personalAccessToken !== undefined
  );
}

export function isGithubBindingDefinition(
  binding?: AlmBindingDefinition & { appId?: string; privateKey?: string; url?: string }
): binding is GithubBindingDefinition {
  return (
    binding !== undefined &&
    binding.appId !== undefined &&
    binding.privateKey !== undefined &&
    binding.url !== undefined
  );
}
