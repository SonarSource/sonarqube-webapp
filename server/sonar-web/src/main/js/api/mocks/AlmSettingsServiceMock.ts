/*
 * SonarQube
 * Copyright (C) 2009-2022 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { cloneDeep } from 'lodash';
import { mockAlmSettingsInstance } from '../../helpers/mocks/alm-settings';
import { AlmKeys, AlmSettingsInstance } from '../../types/alm-settings';
import { getAlmSettings } from '../alm-settings';

export default class AlmSettingsServiceMock {
  almSettings: AlmSettingsInstance[];
  defaultSetting: AlmSettingsInstance[] = [
    mockAlmSettingsInstance({ key: 'conf-final-1', alm: AlmKeys.GitLab }),
    mockAlmSettingsInstance({ key: 'conf-final-2', alm: AlmKeys.GitLab }),
    mockAlmSettingsInstance({ key: 'conf-github-1', alm: AlmKeys.GitHub, url: 'url' }),
    mockAlmSettingsInstance({ key: 'conf-github-2', alm: AlmKeys.GitHub, url: 'url' }),
  ];

  constructor() {
    this.almSettings = cloneDeep(this.defaultSetting);
    (getAlmSettings as jest.Mock).mockImplementation(this.getAlmSettingsHandler);
  }

  getAlmSettingsHandler = () => {
    return Promise.resolve(this.almSettings);
  };

  reset = () => {
    this.almSettings = cloneDeep(this.defaultSetting);
  };
}
