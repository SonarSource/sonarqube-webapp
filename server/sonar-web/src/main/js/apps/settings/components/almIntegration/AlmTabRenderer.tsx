/*
 * SonarQube
 * Copyright (C) 2009-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import { FormattedMessage } from 'react-intl';
import { Link } from 'react-router';
import { Button } from 'sonar-ui-common/components/controls/buttons';
import Tooltip from 'sonar-ui-common/components/controls/Tooltip';
import DeferredSpinner from 'sonar-ui-common/components/ui/DeferredSpinner';
import { translate } from 'sonar-ui-common/helpers/l10n';
import { getEdition, getEditionUrl } from '../../../../helpers/editions';
import {
  AlmBindingDefinition,
  AlmKeys,
  AlmSettingsBindingStatus
} from '../../../../types/alm-settings';
import { EditionKey } from '../../../../types/editions';
import AlmBindingDefinitionBox from './AlmBindingDefinitionBox';
import AlmBindingDefinitionForm, {
  AlmBindingDefinitionFormChildrenProps
} from './AlmBindingDefinitionForm';

export interface AlmTabRendererProps<B> {
  alm: AlmKeys;
  definitionStatus: T.Dict<AlmSettingsBindingStatus>;
  editedDefinition?: B;
  defaultBinding: B;
  definitions: B[];
  form: (props: AlmBindingDefinitionFormChildrenProps<B>) => React.ReactNode;
  help?: React.ReactNode;
  loadingAlmDefinitions: boolean;
  loadingProjectCount: boolean;
  multipleAlmEnabled: boolean;
  onCancel: () => void;
  onCheck: (definitionKey: string) => void;
  onCreate: () => void;
  onDelete: (definitionKey: string) => void;
  onEdit: (definitionKey: string) => void;
  onSubmit: (config: B, originalKey: string) => void;
  optionalFields?: Array<keyof B>;
  submitting: boolean;
  success: boolean;
}

const renderDefaultHelp = (alm: AlmKeys) => (
  <FormattedMessage
    defaultMessage={translate(`settings.almintegration.${alm}.info`)}
    id={`settings.almintegration.${alm}.info`}
    values={{
      link: (
        <Link target="_blank" to="/documentation/analysis/pr-decoration/">
          {translate('learn_more')}
        </Link>
      )
    }}
  />
);

export default function AlmTabRenderer<B extends AlmBindingDefinition>(
  props: AlmTabRendererProps<B>
) {
  const {
    alm,
    definitions,
    definitionStatus,
    editedDefinition,
    form,
    loadingAlmDefinitions,
    loadingProjectCount,
    multipleAlmEnabled,
    optionalFields,
    help = renderDefaultHelp(alm)
  } = props;

  const preventCreation = loadingProjectCount || (!multipleAlmEnabled && definitions.length > 0);
  const creationTooltip = preventCreation ? (
    <FormattedMessage
      id="settings.almintegration.create.tooltip"
      defaultMessage={translate('settings.almintegration.create.tooltip')}
      values={{
        link: (
          <a
            href={getEditionUrl(getEdition(EditionKey.enterprise), {
              sourceEdition: EditionKey.developer
            })}
            rel="noopener noreferrer"
            target="_blank">
            {translate('settings.almintegration.create.tooltip.link')}
          </a>
        ),
        alm: translate('alm', alm)
      }}
    />
  ) : null;

  return (
    <div className="big-padded">
      <DeferredSpinner loading={loadingAlmDefinitions}>
        {definitions.length === 0 && (
          <p className="spacer-top">{translate('settings.almintegration.empty', alm)}</p>
        )}

        <div className={definitions.length > 0 ? 'spacer-bottom text-right' : 'big-spacer-top'}>
          <Tooltip overlay={creationTooltip} mouseLeaveDelay={0.25}>
            <Button
              data-test="settings__alm-create"
              disabled={preventCreation}
              onClick={props.onCreate}>
              {translate('settings.almintegration.create')}
            </Button>
          </Tooltip>
        </div>
        {definitions.map(def => (
          <AlmBindingDefinitionBox
            alm={alm}
            definition={def}
            key={def.key}
            multipleDefinitions={definitions.length > 1}
            onCheck={props.onCheck}
            onDelete={props.onDelete}
            onEdit={props.onEdit}
            status={definitionStatus[def.key]}
          />
        ))}

        {editedDefinition && (
          <AlmBindingDefinitionForm
            bindingDefinition={editedDefinition}
            help={help}
            isSecondInstance={definitions.length === 1}
            onCancel={props.onCancel}
            onSubmit={props.onSubmit}
            optionalFields={optionalFields}>
            {form}
          </AlmBindingDefinitionForm>
        )}
      </DeferredSpinner>
    </div>
  );
}
