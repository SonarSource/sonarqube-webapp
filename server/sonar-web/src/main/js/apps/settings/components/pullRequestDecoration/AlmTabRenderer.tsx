/*
 * SonarQube
 * Copyright (C) 2009-2019 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import { FormattedMessage } from 'react-intl';
import { Link } from 'react-router';
import DeferredSpinner from 'sonar-ui-common/components/ui/DeferredSpinner';
import { translate } from 'sonar-ui-common/helpers/l10n';
import { AlmSettingsBinding, ALM_KEYS } from '../../../../types/alm-settings';
import AlmPRDecorationForm, { AlmPRDecorationFormChildrenProps } from './AlmPRDecorationForm';
import AlmPRDecorationTable from './AlmPRDecorationTable';

export interface AlmTabRendererProps<B> {
  additionalColumnsHeaders: string[];
  additionalColumnsKeys: Array<keyof B>;
  alm: ALM_KEYS;
  editedDefinition?: B;
  defaultBinding: B;
  definitions: B[];
  form: (props: AlmPRDecorationFormChildrenProps<B>) => React.ReactNode;
  loading: boolean;
  multipleAlmEnabled: boolean;
  onCancel: () => void;
  onCreate: () => void;
  onDelete: (definitionKey: string) => void;
  onEdit: (definitionKey: string) => void;
  onSubmit: (config: B, originalKey: string) => void;
  success: boolean;
}

export default function AlmTabRenderer<B extends AlmSettingsBinding>(
  props: AlmTabRendererProps<B>
) {
  const {
    additionalColumnsHeaders,
    additionalColumnsKeys,
    alm,
    defaultBinding,
    definitions,
    editedDefinition,
    form,
    loading,
    multipleAlmEnabled,
    success
  } = props;

  let definition: B | undefined;
  let mappedDefinitions: Array<{ key: string; additionalColumns: string[] }> = [];
  let showEdit: boolean | undefined;

  if (!multipleAlmEnabled) {
    definition = editedDefinition;
    if (definition === undefined && definitions.length > 0) {
      definition = definitions[0];
    }
    showEdit = definition && editedDefinition === undefined;
  } else {
    mappedDefinitions = definitions.map(({ key, ...properties }) => {
      const additionalColumns = additionalColumnsKeys.map(k => (properties as any)[k]);
      return {
        key,
        additionalColumns
      };
    });
  }

  const help = (
    <FormattedMessage
      defaultMessage={translate(`settings.pr_decoration.${alm}.info`)}
      id={`settings.pr_decoration.${alm}.info`}
      values={{
        link: (
          <Link target="_blank" to="/documentation/analysis/pr-decoration/">
            {translate('learn_more')}
          </Link>
        )
      }}
    />
  );

  return multipleAlmEnabled ? (
    <DeferredSpinner loading={loading}>
      <AlmPRDecorationTable
        additionalColumnsHeaders={additionalColumnsHeaders}
        alm={alm}
        definitions={mappedDefinitions}
        onCreate={props.onCreate}
        onDelete={props.onDelete}
        onEdit={props.onEdit}
      />

      {editedDefinition && (
        <AlmPRDecorationForm
          alm={alm}
          bindingDefinition={editedDefinition}
          help={help}
          onCancel={props.onCancel}
          onSubmit={props.onSubmit}
          showInModal={true}>
          {form}
        </AlmPRDecorationForm>
      )}
    </DeferredSpinner>
  ) : (
    <AlmPRDecorationForm
      alm={alm}
      bindingDefinition={definition || defaultBinding}
      help={help}
      hideKeyField={true}
      loading={loading}
      onCancel={props.onCancel}
      onDelete={definition ? props.onDelete : undefined}
      onEdit={showEdit ? props.onEdit : undefined}
      onSubmit={props.onSubmit}
      readOnly={showEdit}
      success={success}>
      {form}
    </AlmPRDecorationForm>
  );
}
