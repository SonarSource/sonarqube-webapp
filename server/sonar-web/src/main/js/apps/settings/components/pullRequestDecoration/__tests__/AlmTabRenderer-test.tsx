/*
 * SonarQube
 * Copyright (C) 2009-2019 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { shallow } from 'enzyme';
import * as React from 'react';
import { mockGithubDefinition } from '../../../../../helpers/mocks/alm-settings';
import { ALM_KEYS, GithubBindingDefinition } from '../../../../../types/alm-settings';
import AlmTabRenderer, { AlmTabRendererProps } from '../AlmTabRenderer';

it('should render correctly for multi-ALM binding', () => {
  expect(shallowRender({ loading: true })).toMatchSnapshot('loading');
  expect(shallowRender()).toMatchSnapshot('loaded');
  expect(shallowRender({ editedDefinition: mockGithubDefinition() })).toMatchSnapshot(
    'editing a definition'
  );
});

it('should render correctly for single-ALM binding', () => {
  expect(shallowRender({ loading: true, multipleAlmEnabled: false })).toMatchSnapshot();
  expect(shallowRender({ multipleAlmEnabled: false })).toMatchSnapshot();
  expect(
    shallowRender({ definitions: [mockGithubDefinition()], multipleAlmEnabled: false })
  ).toMatchSnapshot();
});

function shallowRender(props: Partial<AlmTabRendererProps<GithubBindingDefinition>> = {}) {
  return shallow(
    <AlmTabRenderer
      additionalColumnsHeaders={['url', 'app_id']}
      additionalColumnsKeys={['url', 'appId']}
      alm={ALM_KEYS.GITHUB}
      defaultBinding={mockGithubDefinition()}
      definitions={[mockGithubDefinition()]}
      form={jest.fn()}
      loading={false}
      multipleAlmEnabled={true}
      onCancel={jest.fn()}
      onCreate={jest.fn()}
      onDelete={jest.fn()}
      onEdit={jest.fn()}
      onSubmit={jest.fn()}
      success={false}
      {...props}
    />
  );
}
