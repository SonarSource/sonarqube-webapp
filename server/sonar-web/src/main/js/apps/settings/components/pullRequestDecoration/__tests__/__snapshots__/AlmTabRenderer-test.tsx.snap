// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly for multi-ALM binding: editing a definition 1`] = `
<DeferredSpinner
  loading={false}
  timeout={100}
>
  <AlmPRDecorationTable
    additionalColumnsHeaders={
      Array [
        "url",
        "app_id",
      ]
    }
    alm="github"
    definitions={
      Array [
        Object {
          "additionalColumns": Array [
            "http://github.enterprise.com",
            "123456",
          ],
          "key": "key",
        },
      ]
    }
    onCreate={[MockFunction]}
    onDelete={[MockFunction]}
    onEdit={[MockFunction]}
  />
  <AlmPRDecorationForm
    alm="github"
    bindingDefinition={
      Object {
        "appId": "123456",
        "key": "key",
        "privateKey": "asdf1234",
        "url": "http://github.enterprise.com",
      }
    }
    help={
      <FormattedMessage
        defaultMessage="settings.pr_decoration.github.info"
        id="settings.pr_decoration.github.info"
        values={
          Object {
            "link": <Link
              onlyActiveOnIndex={false}
              style={Object {}}
              target="_blank"
              to="/documentation/analysis/pr-decoration/"
            >
              learn_more
            </Link>,
          }
        }
      />
    }
    onCancel={[MockFunction]}
    onSubmit={[MockFunction]}
    showInModal={true}
  >
    <Component />
  </AlmPRDecorationForm>
</DeferredSpinner>
`;

exports[`should render correctly for multi-ALM binding: loaded 1`] = `
<DeferredSpinner
  loading={false}
  timeout={100}
>
  <AlmPRDecorationTable
    additionalColumnsHeaders={
      Array [
        "url",
        "app_id",
      ]
    }
    alm="github"
    definitions={
      Array [
        Object {
          "additionalColumns": Array [
            "http://github.enterprise.com",
            "123456",
          ],
          "key": "key",
        },
      ]
    }
    onCreate={[MockFunction]}
    onDelete={[MockFunction]}
    onEdit={[MockFunction]}
  />
</DeferredSpinner>
`;

exports[`should render correctly for multi-ALM binding: loading 1`] = `
<DeferredSpinner
  loading={true}
  timeout={100}
>
  <AlmPRDecorationTable
    additionalColumnsHeaders={
      Array [
        "url",
        "app_id",
      ]
    }
    alm="github"
    definitions={
      Array [
        Object {
          "additionalColumns": Array [
            "http://github.enterprise.com",
            "123456",
          ],
          "key": "key",
        },
      ]
    }
    onCreate={[MockFunction]}
    onDelete={[MockFunction]}
    onEdit={[MockFunction]}
  />
</DeferredSpinner>
`;

exports[`should render correctly for single-ALM binding 1`] = `
<AlmPRDecorationForm
  alm="github"
  bindingDefinition={
    Object {
      "appId": "123456",
      "key": "key",
      "privateKey": "asdf1234",
      "url": "http://github.enterprise.com",
    }
  }
  help={
    <FormattedMessage
      defaultMessage="settings.pr_decoration.github.info"
      id="settings.pr_decoration.github.info"
      values={
        Object {
          "link": <Link
            onlyActiveOnIndex={false}
            style={Object {}}
            target="_blank"
            to="/documentation/analysis/pr-decoration/"
          >
            learn_more
          </Link>,
        }
      }
    />
  }
  hideKeyField={true}
  loading={true}
  onCancel={[MockFunction]}
  onDelete={[MockFunction]}
  onEdit={[MockFunction]}
  onSubmit={[MockFunction]}
  readOnly={true}
  success={false}
>
  <Component />
</AlmPRDecorationForm>
`;

exports[`should render correctly for single-ALM binding 2`] = `
<AlmPRDecorationForm
  alm="github"
  bindingDefinition={
    Object {
      "appId": "123456",
      "key": "key",
      "privateKey": "asdf1234",
      "url": "http://github.enterprise.com",
    }
  }
  help={
    <FormattedMessage
      defaultMessage="settings.pr_decoration.github.info"
      id="settings.pr_decoration.github.info"
      values={
        Object {
          "link": <Link
            onlyActiveOnIndex={false}
            style={Object {}}
            target="_blank"
            to="/documentation/analysis/pr-decoration/"
          >
            learn_more
          </Link>,
        }
      }
    />
  }
  hideKeyField={true}
  loading={false}
  onCancel={[MockFunction]}
  onDelete={[MockFunction]}
  onEdit={[MockFunction]}
  onSubmit={[MockFunction]}
  readOnly={true}
  success={false}
>
  <Component />
</AlmPRDecorationForm>
`;

exports[`should render correctly for single-ALM binding 3`] = `
<AlmPRDecorationForm
  alm="github"
  bindingDefinition={
    Object {
      "appId": "123456",
      "key": "key",
      "privateKey": "asdf1234",
      "url": "http://github.enterprise.com",
    }
  }
  help={
    <FormattedMessage
      defaultMessage="settings.pr_decoration.github.info"
      id="settings.pr_decoration.github.info"
      values={
        Object {
          "link": <Link
            onlyActiveOnIndex={false}
            style={Object {}}
            target="_blank"
            to="/documentation/analysis/pr-decoration/"
          >
            learn_more
          </Link>,
        }
      }
    />
  }
  hideKeyField={true}
  loading={false}
  onCancel={[MockFunction]}
  onDelete={[MockFunction]}
  onEdit={[MockFunction]}
  onSubmit={[MockFunction]}
  readOnly={true}
  success={false}
>
  <Component />
</AlmPRDecorationForm>
`;
